In [1]:
# Dependencies
#Importing the dependencies needed to pull API and read APIs
import requests
import json
#Importing the pandas to create DataFrame
import pandas as pd

In [2]:
#Building API query from base url
base_url = "http://api.worldbank.org/v2/"
url=base_url+"lendingtypes"+"?format=json"

In [3]:
#Pulling a response from API query
response=requests.get(url).json()
response

[{'page': '1', 'pages': '1', 'per_page': '50', 'total': '4'},
 [{'id': 'IBD', 'iso2code': 'XF', 'value': 'IBRD'},
  {'id': 'IDB', 'iso2code': 'XH', 'value': 'Blend'},
  {'id': 'IDX', 'iso2code': 'XI', 'value': 'IDA'},
  {'id': 'LNX', 'iso2code': 'XX', 'value': 'Not classified'}]]

From the response above, we can see that there are only 4 results on one page of data

In [4]:
# Get the list of lending types the world bank has
#Selecting index 1 to access second item in the list
lending_type_ids=response[1]
lending_type_ids

[{'id': 'IBD', 'iso2code': 'XF', 'value': 'IBRD'},
 {'id': 'IDB', 'iso2code': 'XH', 'value': 'Blend'},
 {'id': 'IDX', 'iso2code': 'XI', 'value': 'IDA'},
 {'id': 'LNX', 'iso2code': 'XX', 'value': 'Not classified'}]

In [5]:
#Creating empty lists to hold information on the description of the lending type code and the lending type codes
code_description=[]
lending_types=[]

In [6]:
#Looping through each dictionary in the list lending_type_ids and extracting id and value keys
for lending_id in lending_type_ids:
    name=lending_id['id']
    description=lending_id['value']
    #Appending name and description to their respective lists 
    lending_types.append(name)
    code_description.append(description)


In [7]:
#Printing results of lists lending_types and code_description
print(lending_types)
print(code_description)

['IBD', 'IDB', 'IDX', 'LNX']
['IBRD', 'Blend', 'IDA', 'Not classified']


In [8]:
# Determining how many countries fall into each lending type.
#running a request to invetigate nature or new API call
url_test=f"http://api.worldbank.org/v2/country?page=1&format=json"
test_response=requests.get(url_test).json()
test_response

[{'page': 1, 'pages': 7, 'per_page': '50', 'total': 304},
 [{'id': 'ABW',
   'iso2Code': 'AW',
   'name': 'Aruba',
   'region': {'id': 'LCN',
    'iso2code': 'ZJ',
    'value': 'Latin America & Caribbean '},
   'adminregion': {'id': '', 'iso2code': '', 'value': ''},
   'incomeLevel': {'id': 'HIC', 'iso2code': 'XD', 'value': 'High income'},
   'lendingType': {'id': 'LNX', 'iso2code': 'XX', 'value': 'Not classified'},
   'capitalCity': 'Oranjestad',
   'longitude': '-70.0167',
   'latitude': '12.5167'},
  {'id': 'AFG',
   'iso2Code': 'AF',
   'name': 'Afghanistan',
   'region': {'id': 'SAS', 'iso2code': '8S', 'value': 'South Asia'},
   'adminregion': {'id': 'SAS', 'iso2code': '8S', 'value': 'South Asia'},
   'incomeLevel': {'id': 'LIC', 'iso2code': 'XM', 'value': 'Low income'},
   'lendingType': {'id': 'IDX', 'iso2code': 'XI', 'value': 'IDA'},
   'capitalCity': 'Kabul',
   'longitude': '69.1761',
   'latitude': '34.5228'},
  {'id': 'AFR',
   'iso2Code': 'A9',
   'name': 'Africa',
   'reg

#### Test Notes
Based on first item in "test_response" list : {'page': 1, 'pages': 7, 'per_page': '50', 'total': 304}, there are 304 individual items total with each item representing a country. Each page only shows 50 items max per each page for a total of 7 pages. The request above only pulls a response from page 1 of 7 so the url has to be modified to pull all 7 pages

In [9]:
# Determining how many countries fall into each lending type by pulling url for each of the seven pages
#Creating a list to hold each url
query_list=[]

for i in range(1,8):
    url2=f"http://api.worldbank.org/v2/country?page={i}&format=json"
    #print(url2)
    query_list.append(url2)
query_list

['http://api.worldbank.org/v2/country?page=1&format=json',
 'http://api.worldbank.org/v2/country?page=2&format=json',
 'http://api.worldbank.org/v2/country?page=3&format=json',
 'http://api.worldbank.org/v2/country?page=4&format=json',
 'http://api.worldbank.org/v2/country?page=5&format=json',
 'http://api.worldbank.org/v2/country?page=6&format=json',
 'http://api.worldbank.org/v2/country?page=7&format=json']

In [10]:
#Calling a response from each query url (by index) above in the list "query_list"
#Each URL represents one of the seven pages and page_i calls a response for page i of 7
page_1=requests.get(query_list[0]).json()
page_2=requests.get(query_list[1]).json()
page_3=requests.get(query_list[2]).json()
page_4=requests.get(query_list[3]).json()
page_5=requests.get(query_list[4]).json()
page_6=requests.get(query_list[5]).json()
page_7=requests.get(query_list[6]).json()

In [11]:
#Exploring what one of the requests produces
page_1

[{'page': 1, 'pages': 7, 'per_page': '50', 'total': 304},
 [{'id': 'ABW',
   'iso2Code': 'AW',
   'name': 'Aruba',
   'region': {'id': 'LCN',
    'iso2code': 'ZJ',
    'value': 'Latin America & Caribbean '},
   'adminregion': {'id': '', 'iso2code': '', 'value': ''},
   'incomeLevel': {'id': 'HIC', 'iso2code': 'XD', 'value': 'High income'},
   'lendingType': {'id': 'LNX', 'iso2code': 'XX', 'value': 'Not classified'},
   'capitalCity': 'Oranjestad',
   'longitude': '-70.0167',
   'latitude': '12.5167'},
  {'id': 'AFG',
   'iso2Code': 'AF',
   'name': 'Afghanistan',
   'region': {'id': 'SAS', 'iso2code': '8S', 'value': 'South Asia'},
   'adminregion': {'id': 'SAS', 'iso2code': '8S', 'value': 'South Asia'},
   'incomeLevel': {'id': 'LIC', 'iso2code': 'XM', 'value': 'Low income'},
   'lendingType': {'id': 'IDX', 'iso2code': 'XI', 'value': 'IDA'},
   'capitalCity': 'Kabul',
   'longitude': '69.1761',
   'latitude': '34.5228'},
  {'id': 'AFR',
   'iso2Code': 'A9',
   'name': 'Africa',
   'reg

In [12]:
#Pulling item index 1 from page i's response because that is the list that contains
#the information we need to determine how many countries fall into each lending type
list_1=page_1[1]
list_2=page_2[1]
list_3=page_3[1]
list_4=page_4[1]
list_5=page_5[1]
list_6=page_6[1]
list_7=page_7[1]

In [13]:
#Validating that this list contains the information needed
list_1

[{'id': 'ABW',
  'iso2Code': 'AW',
  'name': 'Aruba',
  'region': {'id': 'LCN',
   'iso2code': 'ZJ',
   'value': 'Latin America & Caribbean '},
  'adminregion': {'id': '', 'iso2code': '', 'value': ''},
  'incomeLevel': {'id': 'HIC', 'iso2code': 'XD', 'value': 'High income'},
  'lendingType': {'id': 'LNX', 'iso2code': 'XX', 'value': 'Not classified'},
  'capitalCity': 'Oranjestad',
  'longitude': '-70.0167',
  'latitude': '12.5167'},
 {'id': 'AFG',
  'iso2Code': 'AF',
  'name': 'Afghanistan',
  'region': {'id': 'SAS', 'iso2code': '8S', 'value': 'South Asia'},
  'adminregion': {'id': 'SAS', 'iso2code': '8S', 'value': 'South Asia'},
  'incomeLevel': {'id': 'LIC', 'iso2code': 'XM', 'value': 'Low income'},
  'lendingType': {'id': 'IDX', 'iso2code': 'XI', 'value': 'IDA'},
  'capitalCity': 'Kabul',
  'longitude': '69.1761',
  'latitude': '34.5228'},
 {'id': 'AFR',
  'iso2Code': 'A9',
  'name': 'Africa',
  'region': {'id': 'NA', 'iso2code': 'NA', 'value': 'Aggregates'},
  'adminregion': {'id':

In [14]:
#pulling out lending type id information page by page and appending it to a list called final_list
final_list=[]
#page_1
lending_list1=[]
for i in list_1:
    ltype=i['lendingType']['id']
    lending_list1.append(ltype)
    l1=lending_list1

for i in l1:
    final_list.append(i)

#page_2
lending_list2=[]
for i in list_2:
    ltype=i['lendingType']['id']
    lending_list2.append(ltype)
    l2=lending_list2

for i in l2:
    final_list.append(i)

#page_3
lending_list3=[]
for i in list_3:
    ltype=i['lendingType']['id']
    lending_list3.append(ltype)
    l3=lending_list3

for i in l3:
    final_list.append(i)

#page_4
lending_list4=[]
for i in list_4:
    ltype=i['lendingType']['id']
    lending_list4.append(ltype)
    l4=lending_list4
    
for i in l4:
    final_list.append(i)

#page_5
lending_list5=[]
for i in list_5:
    ltype=i['lendingType']['id']
    lending_list5.append(ltype)
    l5=lending_list5
    
for i in l5:
    final_list.append(i)

#page_6
lending_list6=[]
for i in list_6:
    ltype=i['lendingType']['id']
    lending_list6.append(ltype)
    l6=lending_list6

for i in l6:
        final_list.append(i)

#page_7
lending_list7=[]
for i in list_7:
    ltype=i['lendingType']['id']
    lending_list7.append(ltype)
    l7=lending_list7
    
for i in l7:
        final_list.append(i)    

In [15]:
#Counting the number of occurences of each lending type and placing this information
# in a dictionary called type_count
type_count=dict((x,final_list.count(x)) for x in final_list)
type_count

{'LNX': 74, 'IDX': 59, '': 86, 'IBD': 68, 'IDB': 17}

In [16]:
#calling lending types from first API calls
lending_types

['IBD', 'IDB', 'IDX', 'LNX']

In [17]:
# Printing the number of countries of each lending type
for i in range(len(lending_types)):
    print(f"The number of countries with the lending type {lending_types[i]} is {type_count[lending_types[i]]}")

The number of countries with the lending type IBD is 68
The number of countries with the lending type IDB is 17
The number of countries with the lending type IDX is 59
The number of countries with the lending type LNX is 74


In [18]:
#Creating a DataFrame that contains the lending type, the description of that lending type code,
#as well as the number of countries with that lending type
num_countries=[]
for i in range(len(lending_types)):
    number=type_count[lending_types[i]]
    num_countries.append(number)
num_countries

[68, 17, 59, 74]

In [19]:
#Creating a DataFrame to show number of countries and code description for each lending type
table=pd.DataFrame(zip(lending_types,num_countries,code_description))
table

0   1               2
0  IBD  68            IBRD
1  IDB  17           Blend
2  IDX  59             IDA
3  LNX  74  Not classified

In [20]:
#Naming each column in the DataFrame to accurately describe contents of the columns
table=table.rename(columns={0:"Lending Type",1:"No. of countries",2:"Description of lending type code"})
table

Lending Type  No. of countries Description of lending type code
0          IBD                68                             IBRD
1          IDB                17                            Blend
2          IDX                59                              IDA
3          LNX                74                   Not classified